### DISCUSSION TITLE
Trying to get a simple example to work using Microsoft.Build, but can't

### DISCUSSION THREAD

**Discussion by awa5114** [dotnet/msbuild#10256] [Q&A] [answered]
I'm trying to use the `Microsoft.Build` package to programmatically probe a third-party project file for some information. I started off with the following simple example:

```cs
using Microsoft.Build.Locator;
using Microsoft.Build.Evaluation;

string projectFilePath = @"Path/To/ThirdPartyApp.csproj";

MSBuildLocator.RegisterDefaults();

// Load an existing project
Project proj = new Project(projectFilePath);

// Access an ItemGroup
ICollection<ProjectItem> itemGroup = proj.GetItems("Compile");

// Print all files matched by the Item
foreach (ProjectItem item in itemGroup)
{
    Console.WriteLine(item.EvaluatedInclude);
}
```

The project file for my application (i.e. the one that's supposed to do the probing, **not** `ThirdPartyApp.csproj`) started off looking like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Build" Version="17.10.4" />
    <PackageReference Include="Microsoft.Build.Locator" Version="1.7.8" />
  </ItemGroup>

</Project>
```

When I ran it, however I got the following error:

> error : A PackageReference to Microsoft.Build.* without ExcludeAssets="runtime" exists in your project. This will cause MSBuild assemblies to be copied to your output directory, causing your application to load them at runtime. To use the copy of MSBuild registered by MSBuildLocator, set ExcludeAssets="runtime" on the MSBuild PackageReferences. To disable this check, set the property DisableMSBuildAssemblyCopyCheck=true in your project file (not recommended as you must distributed all of MSBuild + associated
 toolset).

I then tried to add the `ExcludeAssets="runtime"` property as follows:

```
<PackageReference Include="Microsoft.Build" Version="17.10.4" ExcludeAssets="runtime"/>
```

When I run the app, however I get a runtime error:

> Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
File name: 'Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Program.<Main>$(String[] args)

Is there something fundamentally wrong with how I'm trying to use `Microsoft.Build`? Is this a bug/expected behavior?

**Comment by rainersigwald** [comment]
Can you see if the docs answer your question, specifically this part? https://learn.microsoft.com/visualstudio/msbuild/find-and-use-msbuild-versions?view=vs-2022#register-instance-before-calling-msbuild

Unfortunately we can't do much with the error message for this problem, since it's thrown by the .NET runtime itself.

**Reply by awa5114** [reply]
Thanks for your response. As mentioned in the original question I am already registering the instance using `Microsoft.Build.Locator`. I am therefore familiar with the part of the documentation you linked.

> Unfortunately we can't do much with the error message for this problem, since it's thrown by the .NET runtime itself.

I'm pretty much adding a reference and using the types in `Microsoft.Build` the same way I would any other nuget package. I'm not sure why `Microsoft.Build` would be any different?

**Reply by KalleOlaviNiemitalo** [reply, marked as answer]
@amine-aboufirass, note this text in the documentation:

> You can't reference any MSBuild types (from the Microsoft.Build namespace) in the method that calls MSBuildLocator.

In your code, the `Project proj = new Project(projectFilePath);` statement seems to be in the same method as the `MSBuildLocator.RegisterDefaults()` call.  You should move it to a second method, so that things can proceed in this order:

1. The .NET Runtime compiles the first method.  Because the first method does not reference MSBuild types such as `Project`, the .NET Runtime does not need to find the definitions of those types yet.
2. The first method starts running, and calls MSBuildLocator.RegisterDefaults(), which makes the MSBuild types available.
3. The first method calls the second method.
4. The .NET Runtime compiles the second method.  Because the second method has a local variable of type `Project`, the .NET Runtime needs to find the definition of this type.  Thanks to MSBuildLocator.RegisterDefaults(), the definition is available.
5. The second method starts running, and creates an instance of `Project`.
